### PR TITLE
[Development] Update "New condition" review design

### PR DESCRIPTION
**Pull request by Mottie**
## Description

Updated "New condition" review card design. To accomplish this, a new review `ObjectField` setting was added.

Related:
- Design: https://github.com/department-of-veterans-affairs/va.gov-team/issues/12678
- Slack: https://dsva.slack.com/archives/C5AGLBNRK/p1598984849010700
- Documentation: https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/pull/386

## Testing done

- Unit test added for `ui:objectViewField` setting
- Ran axe-coconut to check a11y compliance

## Screenshots

<details><summary>Before</summary>

Screen Shot 2020-09-01 at 12 44 00 PM</details>

<details><summary>After</summary>

Screen Shot 2020-09-01 at 1 35 34 PM</details>

## Acceptance criteria
- [x] New conditions don't show unnecessary text
- [x] Edit button is aligned with new condition text

## Definition of done
- [ ] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by ndsprinkle — APPROVED** [review]

**Comment by meganhkelley** [comment]
Hi @Mottie thanks for including the ticket and Slack thread for context! My understanding is that: 

* Your team received non-launch-blocking design feedback in Staging Review to alter the New Condition edit button placement 
* Your proposed change requires a change to forms library code 

Some questions for you: 

* What impact does this have on other forms on VA.gov? Will this change affect any live forms? 
* How has the New Condition edit button been handled in the past? (related to above Q)
* Can you confirm that this is not launch blocking for y'all? (That is my understanding based on it being "should" feedback)

**Comment by Mottie** [comment]
Hi @meganhkelley!

> * Your team received non-launch-blocking design feedback in Staging Review to alter the New Condition edit button placement
> * Your proposed change requires a change to forms library code

You are correct in your understanding.

> * What impact does this have on other forms on VA.gov? Will this change affect any live forms?

This change will apply to all other forms on VA.gov, but the developer must add a new option `ui:objectViewField` to the `uiSchema` in their form to utilitize this change.

> * How has the New Condition edit button been handled in the past? (related to above Q)

If you look at the first comment in this PR, you can see a before & after screenshots. The edit button is present in both cases, the location of the edit button has been moved from the header row into the `<dd>` of the review row.

<details><summary>Header row HTML (before screenshot)</summary>


```html
<div class="va-growable-background">
  <div class="row small-collapse">
    <form class="rjsf" novalidate="">
      <div class="form-review-panel-page-header-row">
        <button
          type="button"
          class="edit-btn primary-outline"
          aria-label="Edit New condition"
        >
          Edit
        </button>
      </div>
      <dl class="review">
        <div class="review-row">
          <dt>
            <p>
              If you know the name of your condition, you can type it here. You
              can write whatever you want and we’ll make suggestions for
              possible disabilities. (Shorter descriptions are better. For
              example, foot pain, back pain, or hearing loss.)
            </p>
          </dt>
          <dd><span>asthma</span></dd>
        </div>
      </dl>
      <div></div>
    </form>
  </div>
</div>
```
</details>

<details>
<summary>Changes using 'ui:objectViewField' customization (after screenshot)</summary>


```html
<div class="va-growable-background">
  <div class="row small-collapse">
    <form class="rjsf" novalidate="">
      <dl class="vads-u-width--full">
        <div class="review-row vads-u-display--flex vads-u-justify-content--space-between vads-u-align-items--center">
          <dt class="capitalize-first-letter">
            <span>asthma</span>
          </dt>
          <dd>
            <button
              type="button"
              class="edit-btn primary-outline"
              aria-label="Edit New condition"
            >
              Edit
            </button>
          </dd>
        </div>
      </dl>
      <div></div>
    </form>
  </div>
</div>
```
</details>

> * Can you confirm that this is not launch blocking for y'all? (That is my understanding based on it being "should" feedback)

This change is not launch blocking

**Comment by Mottie** [comment]
I've discovered additional use cases for adding a `ui:objectViewField` field renderer:

- Form 526 includes `view:bankAccount` (for newly added bank info) and `view:originalBankAccount` (for loaded account information). The `ObjectField` does not pass down data from the parent, so the `view:bankAccount` always renders with empty data if `view:originalBankAccount` data is available, but no new data was provided - see https://github.com/department-of-veterans-affairs/va.gov-team/issues/2345
- If you want to render a `<ul>` with `<li>` items, which is what this PR _should_ be doing, instead of `<dl>` with nested `<dt>` and `<dd>` elements.

**Comment by meganhkelley** [comment]
Hi @Mottie ! We've been discussing VSP process related to this one as well, and here's where we've landed: 

- BAM1 is free to use this new design in "your" own app _(you did not ask about this, but I'm clarifying anyway)_
- VSP would like to ask BAM 1 to submit this new design for design system council approval to be used by other teams as well. Could you please create a ticket for this in the "Proposals" column here? https://github.com/department-of-veterans-affairs/vets-design-system-documentation/projects/3 Let me know if you have any Qs about that. 
- Because BAM1's use of this new design in "your" specific app involves a global code change, VSP FE Tools needs some extra time to review this work, and work with the BAM1 team on your code implementation (confirm that this is a 👍 change to the forms library). That global code change can proceed before bullet 2 (design system council review), so @cvalarida is going to take a look at this within the next few days.

Let me know if you — Robin or Chris — have any questions or concerns!

**Comment by Mottie** [comment]
@meganhkelley Please see https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/309

File: src/platform/forms-system/src/js/review/ObjectField.jsx
Comment (cvalarida) [inline src/platform/forms-system/src/js/review/ObjectField.jsx:174]: What do you think of passing an `onEdit` prop instead of the whole `editButton`? As-written, the custom `objectViewField` component can neither change the button (styling, wording, etc.) nor create a new one with that `formContext.onEdit` callback.

File: src/applications/disability-benefits/all-claims/components/ObjectViewField.jsx
Comment (cvalarida) [inline src/applications/disability-benefits/all-claims/components/ObjectViewField.jsx:3]: This is app code, so don't feel any obligation to listen to me on this one, but I had a recommendation anyhow. :smile: 

Consider using a less generic name for this. I find either naming the pattern (`SimpleDisplayReviewField`) or naming the use (`ConditionReviewField`) works pretty well.

**Review by cvalarida — APPROVED** [review]
Sorry for taking so long on this one, but it looks solid. I thought this would introduce more global changes here, but that doesn't look like the right answer. What you've got is great. :+1: :shipit:

File: src/platform/forms-system/src/js/review/ObjectField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/review/ObjectField.jsx:174]: I do like this idea better than what I have in place... I'll work on updating this ASAP.

File: src/applications/disability-benefits/all-claims/components/ObjectViewField.jsx
Comment (Mottie) [reply src/applications/disability-benefits/all-claims/components/ObjectViewField.jsx:3]: I have no opinion on this, but I can't think of a good name. Any preference?

File: src/platform/forms-system/src/js/review/ObjectField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/review/ObjectField.jsx:174]: Hmm, I did pass in the `props`, which contains `formContext.onEdit`; so it's already covered.

I was just looking at this thinking, that the `editButton` contains `editLabel` which is derived from:

```js
const editLabel =
  _.get('ui:options.ariaLabelForEditButtonOnReview', uiSchema) ||
  `Edit ${title}`;
```

so maybe I should include the `editLabel`, then... etc, etc...

So I came to this conclusion... flatten the `this.props` into the passed object, and keep the `editButton`, so you can do whatever you want. I'll update the docs so make it clear that the `editButton` is there, but you can customize it by showing an example.

File: src/applications/disability-benefits/all-claims/components/ObjectViewField.jsx
Comment (cvalarida) [reply src/applications/disability-benefits/all-claims/components/ObjectViewField.jsx:3]: :man_shrugging: I'd probably go with `ConditionReviewField` myself, but it's your call, of course. :smile:

File: src/platform/forms-system/src/js/review/ObjectField.jsx
Comment (cvalarida) [reply src/platform/forms-system/src/js/review/ObjectField.jsx:174]: That makes enough sense. I'm not a superfan of spreading the props to a child component, but that seems appropriate in this scenario.

I guess all I'd like to see now is renaming `editButton` to `defaultEditButton` to make it clear that it's just a default.

If you wanna earn Bonus Points™ (which are, of course, very valuable :stuck_out_tongue_winking_eye:), you could make the default edit button a standalone function component which accepts the `title` and `onEdit` function. That way we're not passing an entire component as a prop, which is kinda gross.

(Martin Fowler gave this refactor a name: "Replace Parameter with Query", page 324, _Refactoring, Second Edition._)

Comment (Mottie) [reply src/platform/forms-system/src/js/review/ObjectField.jsx]: Done! Let me know what you think!

**Comment by Mottie** [comment]
This PR has been updated:
- Will keep the platform addition of the `ui:objectViewField` setting
- Disabled changes made to form 526's `addDisabilities` UI schema that modified the review & submit page rendering (as seen in the screenshots). I will include this in a later PR, once design changes have been approved.

**Pull request merged**
